template/template.go

```diff
@@ -296,3 +308,238 @@ func (e *executer) resolveOptions(options []ast.Option) (map[string]any, error)
+
+		rv := res[i]
+
+		matches := matchSelectorKeys(rv, keys)
+		pref = append(pref, matches)
+	}
+
+	return pref
+}
+
+func (e *executer) filterVariants(m ast.Matcher, pref [][]string) []ast.Variant {
+	// Step 3: Filter Variants
+	var filteredVariants []ast.Variant
+
+	for _, variant := range m.Variants {
+		matchesAllSelectors := true
+
+		for i, keyOrder := range pref {
+			key := variant.Keys[i]
+			if key.String() == "*" {
+				continue
+			}
+
+			var ks string
+
+			switch key := key.(type) {
+			case ast.CatchAllKey:
+				continue
+			case ast.NameLiteral:
+				ks = string(key)
+			case ast.QuotedLiteral:
+				ks = string(key)
+			case ast.NumberLiteral:
+				ks = fmt.Sprint(key)
+			}
+
+			if !slices.Contains(keyOrder, ks) {
+				matchesAllSelectors = false
+				break
+			}
+		}
+
+		if matchesAllSelectors {
+			filteredVariants = append(filteredVariants, variant)
+		}
+	}
+
+	return filteredVariants
+}
+
+func (e *executer) sortVariants(filteredVariants []ast.Variant, pref [][]string) []SortableVariant {
+	// Step 4: Sort Variants
+	sortable := make([]SortableVariant, 0, len(filteredVariants))
+
+	for _, variant := range filteredVariants {
+		sortable = append(sortable, SortableVariant{Score: -1, Variant: variant})
+	}
+
+	for i := len(pref) - 1; i >= 0; i-- {
+		matches := pref[i]
+
+		for tupleIndex, tuple := range sortable {
+			key := tuple.Variant.Keys[i]
+			currentScore := len(matches)
+
+			if key.String() != "*" {
+				var ks string
+
+				switch key := key.(type) {
+				case ast.CatchAllKey:
+					continue
+				case ast.NameLiteral:
+					ks = string(key)
+				case ast.QuotedLiteral:
+					ks = string(key)
+				case ast.NumberLiteral:
+					ks = fmt.Sprint(key)
+				}
+
+				if position := findPosition(ks, matches); position != -1 {
+					currentScore = position
+				}
+			}
+
+			sortable[tupleIndex].Score = currentScore
+		}
+
+		sort.Sort(SortableVariants(sortable))
+	}
+
+	return sortable
+}
+
+func (e *executer) selectBestVariant(sortable []SortableVariant) (string, error) { //nolint:unparam
+	// Select the best variant
+	bestVariant := sortable[0].Variant
+
+	var output string
+
+	for _, patternElement := range bestVariant.QuotedPattern {
+		if err := e.resolvePattern(patternElement); err != nil {
+			return "", fmt.Errorf("resolve pattern element: %w", err)
+		}
+	}
+
+	return output, nil
+}
+
+func matchSelectorKeys(rv any, keys []string) []string {
+	var matches []string
+
+	for _, key := range keys {
+		if value, ok := rv.(string); ok {
+			if key == value {
+				matches = append(matches, key)
+			}
+		}
+	}
+
+	return matches
+}
+
+type SortableVariant struct {
+	Variant ast.Variant
+	Score   int
+}
+
+type SortableVariants []SortableVariant
+
+func (s SortableVariants) Len() int {
+	return len(s)
+}
+
+func (s SortableVariants) Less(i, j int) bool {
+	return s[i].Score < s[j].Score
+}
+
+func (s SortableVariants) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func findPosition(ks string, matches []string) int {
```

Comment:
as you imported slices package, you could use `slices.Index ` instead of this func